api/tests/test_asset_status_api.py

```diff
@@ -0,0 +1,166 @@
+            model_number=self.test_assetmodel1,
+            assigned_to=self.normal_user
+        )
+        self.test_asset.save()
+        self.asset = Asset.objects.get(asset_code="IC001")
+        self.asset_status = AssetStatus.objects.get(asset=self.asset)
+
+        self.asset_status_urls = reverse('asset-status-list')
+
+    def test_non_authenticated_user_view_asset_status(self):
+        response = client.get(self.asset_status_urls)
+        self.assertEqual(response.data, {
+            'detail': 'Authentication credentials were not provided.'
+        })
+
+    @patch('api.authentication.auth.verify_id_token')
+    def test_authenticated_user_view_asset_status(self, mock_verify_id_token):
+        mock_verify_id_token.return_value = {'email': self.normal_user.email}
+        response = client.get(
+            self.asset_status_urls,
+            HTTP_AUTHORIZATION="Token {}".format(self.token_user))
+        self.assertIn(self.asset_status.asset_id, response.data[0].values())
+
+        self.assertEqual(len(response.data), Asset.objects.count())
+        self.assertEqual(response.status_code, 200)
+
+    @patch('api.authentication.auth.verify_id_token')
+    def test_authenticated_user_view_single_asset_status(
+            self, mock_verify_id_token):
+        mock_verify_id_token.return_value = {'email': self.normal_user.email}
+        response = client.get(
+            '{}{}/'.format(self.asset_status_urls, self.asset_status.id),
+            HTTP_AUTHORIZATION="Token {}".format(self.token_user))
+        self.assertIn(self.asset_status.asset_id, response.data.values())
+        self.assertEqual(response.status_code, 200)
+
+    @patch('api.authentication.auth.verify_id_token')
+    def test_authenticated_user_post_asset_status(self, mock_verify_id_token):
+        mock_verify_id_token.return_value = {'email': self.normal_user.email}
+        data = {
+            "asset": self.test_asset.serial_number,
+            "current_status": "Available"
+        }
+        response = client.post(
+            self.asset_status_urls,
+            data=data,
+            HTTP_AUTHORIZATION="Token {}".format(self.token_user))
+        self.assertIn(data['asset'], response.data.values())
+
+        self.assertEqual(response.status_code, 201)
+
+    @patch('api.authentication.auth.verify_id_token')
+    def test_authenticated_user_post_invalid_serial_number(
+            self, mock_verify_id_token):
+        mock_verify_id_token.return_value = {'email': self.normal_user.email}
+        data = {
+            "asset": "Invalid",
+            "current_status": "Available"
+        }
+        response = client.post(
+            self.asset_status_urls,
+            data=data,
+            HTTP_AUTHORIZATION="Token {}".format(self.token_user))
+        self.assertEqual(response.data, {
+            'asset': ['Object with serial_number=Invalid does not exist.']
+        })
+
+        self.assertEqual(response.status_code, 400)
```

Comment:
This is okay but how come the status code is not 404 ?

Reply:
Its supposed to be 400 not 404 because the server cannot process the request due to something that is perceived to be a client error which is true.